### PR TITLE
[FIX JENKINS-20520] Save Jenkins after adding global user permission

### PR DESCRIPTION
**Pull request by daniel-beck**
[JENKINS-20520](https://issues.jenkins-ci.org/browse/JENKINS-20520)

@jglick

### CODE REVIEW

File: src/test/java/org/jenkinsci/plugins/matrixauth/PermissionAdderTest.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+public class PermissionAdderTest {
+
+    @Rule
+    public RestartableJenkinsRule r = new RestartableJenkinsRule();
+
+    @Test
+    @Issue("JENKINS-20520")
+    public void ensureSavingAfterInitialUser() {
+        r.addStep(new Statement() {
+            @Override
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/matrixauth/PermissionAdderTest.java]: BTW once you are on 2.60.x+ you can make this a lot prettier with lambdas.

File: src/test/java/org/jenkinsci/plugins/matrixauth/PermissionAdderTest.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            public void evaluate() throws Throwable {
+                r.j.jenkins.setSecurityRealm(new HudsonPrivateSecurityRealm(true));
+                r.j.jenkins.setAuthorizationStrategy(new GlobalMatrixAuthorizationStrategy());
+                r.j.jenkins.save();
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/matrixauth/PermissionAdderTest.java]: BTW in 2.51+ this is unnecessary: https://github.com/jenkinsci/jenkins/pull/2790

File: src/test/java/org/jenkinsci/plugins/matrixauth/PermissionAdderTest.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
+package org.jenkinsci.plugins.matrixauth;
+
+import com.gargoylesoftware.htmlunit.html.HtmlPage;
+import hudson.model.User;
+import hudson.security.ACL;
+import hudson.security.ACLContext;
+import hudson.security.GlobalMatrixAuthorizationStrategy;
+import hudson.security.HudsonPrivateSecurityRealm;
+import hudson.security.pages.SignupPage;
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/matrixauth/PermissionAdderTest.java]: Wow, never heard of this before. Once the private realm gets moved into a plugin (?), I guess this should go into a `test-jar`.

File: src/test/java/org/jenkinsci/plugins/matrixauth/PermissionAdderTest.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+                JenkinsRule.WebClient wc = r.j.createWebClient();
+                SignupPage signup = new SignupPage(wc.goTo("signup"));
+                signup.enterUsername("alice");
+                signup.enterPassword("alice");
+                signup.enterFullName("Alice User");
+                HtmlPage success = signup.submit(r.j);
+
+                try (ACLContext _ = ACL.as(User.get("alice"))) {
+                    Assert.assertTrue(r.j.jenkins.getACL().hasPermission(Jenkins.ADMINISTER));
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/matrixauth/PermissionAdderTest.java]: More simply:

```java
assertTrue(r.j.jenkins.hasPermission(Jenkins.ADMINISTER));
```

More simply still, delete the `try` block and:

```java
assertTrue(r.j.jenkins.getACL().hasPermission(User.get("alice").impersonate(), Jenkins.ADMINISTER));
```

Comment (daniel-beck) [reply src/test/java/org/jenkinsci/plugins/matrixauth/PermissionAdderTest.java]: Thanks, used this in https://github.com/jenkinsci/matrix-auth-plugin/pull/22 but forgot to update it here when I remembered how this works.

File: src/test/java/org/jenkinsci/plugins/matrixauth/PermissionAdderTest.java (hunk 2 above)
Comment (daniel-beck) [reply src/test/java/org/jenkinsci/plugins/matrixauth/PermissionAdderTest.java]: I want this to be explicit.

File: src/test/java/org/jenkinsci/plugins/matrixauth/PermissionAdderTest.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+                r.j.jenkins.save();
+
+                JenkinsRule.WebClient wc = r.j.createWebClient();
+                SignupPage signup = new SignupPage(wc.goTo("signup"));
+                signup.enterUsername("alice");
+                signup.enterPassword("alice");
+                signup.enterFullName("Alice User");
+                HtmlPage success = signup.submit(r.j);
+
+                Assert.assertTrue(r.j.jenkins.getACL().hasPermission(User.get("alice").impersonate(), Jenkins.ADMINISTER));
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/matrixauth/PermissionAdderTest.java:41]: Failing in https://github.com/jenkinsci/bom/pull/359 for reasons I am still struggling to understand: `User.get("alice", false, Collections.emptyMap())` returns null. Only when `jth.jenkins-war.path` is set to the megawar.

**Pull request merged**